### PR TITLE
Handle delete history branch

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
Handle delete history branch when mutable state does not exist

**Why?**
Handle two cases:
1. ms is gone. We should try to delete the history branch.
2. version check failed. We should try to delete the history branch.

With this, the task can be retried and only delete the leftover history data.

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: service/history/timerQueueActiveTaskExecutor.go [hunk 1]
```diff
@@ -447,8 +447,8 @@ func (t *timerQueueActiveTaskExecutor) executeActivityRetryTimerTask(
 		}
 		return nil
 	}
-	ok, err = verifyTaskVersion(t.shard, t.logger, namespace.ID(task.NamespaceID), activityInfo.Version, task.Version, task)
-	if err != nil || !ok {
+	ok = verifyTaskVersion(t.shard, t.logger, mutableState.GetNamespaceEntry(), activityInfo.Version, task.Version, task)
+	if !ok {
 		return err
```
Comment (yiminc) [inline service/history/timerQueueActiveTaskExecutor.go:452]: return err will return nil, if you intend to return nil, then make it clear by return nil

File: service/history/timerQueueTaskExecutorBase.go [hunk 2]
```diff
@@ -91,17 +95,28 @@ func (t *timerQueueTaskExecutorBase) executeDeleteHistoryEventTask(
 	defer func() { release(retError) }()
 
 	mutableState, err := loadMutableStateForTimerTask(weContext, task, t.metricsClient, t.logger)
-	if err != nil {
+	switch err.(type) {
+	case nil:
+		// continue the following step
+	case *serviceerror.NotFound:
+		return t.deleteHistoryBranch(task.BranchToken)
```
Comment (yiminc) [inline service/history/timerQueueTaskExecutorBase.go:102]: add some comment to explain the case

File: service/history/timerQueueTaskExecutorBase.go [hunk 3]
```diff
@@ -91,17 +95,28 @@ func (t *timerQueueTaskExecutorBase) executeDeleteHistoryEventTask(
...
 
 	lastWriteVersion, err := mutableState.GetLastWriteVersion()
 	if err != nil {
 		return err
 	}
-	ok, err := verifyTaskVersion(t.shard, t.logger, namespace.ID(task.NamespaceID), lastWriteVersion, task.Version, task)
-	if err != nil || !ok {
-		return err
+	if ok := verifyTaskVersion(t.shard, t.logger, mutableState.GetNamespaceEntry(), lastWriteVersion, task.Version, task); !ok {
+		currentBranchToken, err := mutableState.GetCurrentBranchToken()
```
Comment (yiminc) [inline service/history/timerQueueTaskExecutorBase.go:112]: add comments to explain this case

File: service/history/timerQueueTaskExecutorBase.go [hunk 4]
```diff
@@ -91,17 +95,28 @@ func (t *timerQueueTaskExecutorBase) executeDeleteHistoryEventTask(
...
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(task.BranchToken, currentBranchToken) {
+			return t.deleteHistoryBranch(task.BranchToken)
+		}
+		return nil
```
Comment (yiminc) [inline service/history/timerQueueTaskExecutorBase.go:119]: add comment for this case as well. this means version changed, but branch token keeps the same, we will do nothing here and wait for the new task (for new version) to delete history.

File: service/history/timerQueueTaskExecutorBase.go [hunk 5]
```diff
@@ -122,3 +137,14 @@ func (t *timerQueueTaskExecutorBase) getNamespaceIDAndWorkflowExecution(
 		RunId:      task.GetRunID(),
 	}
 }
+
+func (t *timerQueueTaskExecutorBase) deleteHistoryBranch(branchToken []byte) error {
+	if len(branchToken) > 0 {
+		// only delete the history if mutable state is gone
```
Comment (yiminc) [inline service/history/timerQueueTaskExecutorBase.go:143]: this comment is out of context here.
here you just delete the given branch, or does it actually check mutable state again?

File: service/history/timerQueueTaskExecutorBase.go [hunk 6]
```diff
@@ -91,17 +95,34 @@ func (t *timerQueueTaskExecutorBase) executeDeleteHistoryEventTask(
...
+			return err
+		}
+		// the mutable state has a newer version and the branch token is updated
+		// use task branch token to delete the original branch
+		if !bytes.Equal(task.BranchToken, currentBranchToken) {
+			return t.deleteHistoryBranch(task.BranchToken)
+		}
+		// it is not expected to hit here. Usually, an updated version means a different branch token
+		// this is to prevent us from deleting the correct branch token
+		return nil
```
Comment (yycptt) [inline service/history/timerQueueTaskExecutorBase.go:125]: if this case is not expected and can be a potential bug, probably emit some log or metric so we know it happens?

**Review by yycptt — APPROVED** [review]

**Pull request merged**